server/src/main/java/com/cloud/vm/schedule/VMScheduleManagerImpl.java

```diff
@@ -0,0 +1,598 @@
+
+        vmSchedule.setState(VMSchedule.State.DISABLED);
+        return vmScheduleDao.update(vmSchedule.getId(), vmSchedule);
+    }
+
+    @Override
+    @ActionEvent(eventType = EventTypes.EVENT_VMSCHEDULE_UPDATE, eventDescription = "update VM Schedule")
+    public VMSchedule updateVMSchedule(UpdateVMScheduleCmd cmd) {
+        VMScheduleVO vmSchedule = vmScheduleDao.findById(cmd.getId());
+        if (vmSchedule == null) {
+            throw new InvalidParameterValueException("unable to find the vm schedule with id " + cmd.getId());
+        }
+        String description = cmd.getDescription();
+        String schedule = cmd.getSchedule();
+        String intervalType = cmd.getIntervalType();
+        String action = cmd.getAction();
+        String tag = cmd.getTag();
+        String timezone = cmd.getTimezone();
+
+        if (vmSchedule.getState() == VMSchedule.State.DISABLED) {
+            if (description != null)
+                vmSchedule.setDescription(description);
+            if(schedule != null)
+                vmSchedule.setSchedule(schedule);
+            if(intervalType != null)
+                vmSchedule.setScheduleType(intervalType);
+            if (action != null)
+                vmSchedule.setAction(action);
+            if (tag != null)
+                vmSchedule.setTag(tag);
+            if (timezone != null)
+                vmSchedule.setTimezone(timezone);
+
+            vmScheduleDao.update(vmSchedule.getId(), vmSchedule);
+        } else {
+            throw new InvalidParameterValueException("Disable the state of VM Schedule before updating it");
+        }
+        return vmSchedule;
+    }
+
+    public void poll(final Date timestamp) {
+        currentTimestamp = timestamp;
+        GlobalLock scanLock = GlobalLock.getInternLock("vm.poll");
+
+        try {
+            if (scanLock.lock(5)) {
+                try {
+                    checkStatusOfCurrentlyExecutingVMs();
+                } finally {
+                    scanLock.unlock();
+                }
+            }
+        } finally {
+            scanLock.releaseRef();
+        }
+
+        scanLock = GlobalLock.getInternLock("vm.poll");
+        try {
+            if (scanLock.lock(5)) {
+                try {
+                    scheduleVMs();
+                } finally {
+                    scanLock.unlock();
+                }
+            }
+        } finally {
+            scanLock.releaseRef();
+        }
+    }
+
+    @DB
+    public void scheduleVMs() {
+        String displayTime = DateUtil.displayDateInTimezone(DateUtil.GMT_TIMEZONE, currentTimestamp);
+        LOGGER.debug("VM poll is being called at " + displayTime);
+
+        final List<VMScheduleVO> vmsToBeExecuted = vmScheduleDao.listAll();
+        for (final VMScheduleVO vmSchedule : vmsToBeExecuted) {
+            final long timeDifference = DateUtil.getTimeDifference(vmSchedule.getScheduledTimestamp(), currentTimestamp);
+
+            if (timeDifference <= 1) {
```

Comment:
@rahulbcn27 
can you explain what this check does `if (timeDifference <= 1)`?
should `VMSchedulerInterval.value()` be used here ?